File: Client/Assets/Script/Packet/PacketHeader.cs
using System;

class PacketHeader {
    public int packetID;
    public int length;

    public static readonly int Size = 8;

    public PacketHeader()
    {
        Init();
    }

    public void Init()
    {
        packetID = -1;
        length = -1;
    }

    public byte[] Serialize()
    {
        var data = new byte[Size];
        Buffer.BlockCopy(BitConverter.GetBytes(packetID), 0, data, 0, 4);
        Buffer.BlockCopy(BitConverter.GetBytes(length), 0, data, 4, 4);

        return data;
    }

    public void Deserialize(byte[] data)
    {
        packetID = BitConverter.ToInt32(data, 0);
        length = BitConverter.ToInt32(data, 4);
    }
}

File: Server/Packet/src/PacketHeader.cpp
#include "stdafx.h"
#include "PacketHeader.h"

namespace Packet
{
	Header::Header()
	{
		Init();
	}

	void Header::Init()
	{
		packetID = -1;
		length = -1;
	}

	Header::Buffer Header::Serialize()
	{
		Buffer buffer;
		buffer.data = new char[Size];

		memcpy(buffer.data, &packetID, sizeof(packetID));
		memcpy(buffer.data + sizeof(packetID), &length, sizeof(length));

		return buffer;
	}

	void Header::Serialize(void* buffer)
	{
		memcpy(buffer, &packetID, sizeof(packetID));
		memcpy(static_cast<char*>(buffer) + sizeof(packetID), &length, sizeof(length));
	}

	void Header::Deserialize(Buffer buffer)
	{
		Deserialize(buffer.data);
	}

	void Header::Deserialize(char* data)
	{
		memcpy(&packetID, data, sizeof(packetID));
		memcpy(&length, data + sizeof(packetID), sizeof(length));
	}
}